src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/ProjectRuleHandler.cs

```diff
@@ -37,19 +37,35 @@ internal class ProjectRuleHandler : DependenciesRuleHandlerBase
                                   IDependenciesSnapshotProvider snapshotProvider,
                                   IUnconfiguredProjectCommonServices commonServices)
         {
-            AggregateSnapshotProvider = aggregateSnapshotProvider;
             SnapshotProvider = snapshotProvider;
-            CommonServices = commonServices;
 
-            AggregateSnapshotProvider.SnapshotChanged += OnSnapshotChanged;
-            AggregateSnapshotProvider.SnapshotProviderUnloading += OnSnapshotProviderUnloading;
+            aggregateSnapshotProvider.SnapshotChanged += OnAggregateSnapshotChanged;
+            aggregateSnapshotProvider.SnapshotProviderUnloading += OnAggregateSnapshotProviderUnloading;
 
-            CommonServices.Project.ProjectUnloading += OnUnconfiguredProjectUnloading;
+            // Unregister event handlers when the project unloads
+            commonServices.Project.ProjectUnloading += OnUnconfiguredProjectUnloading;
+
+            Task OnUnconfiguredProjectUnloading(object sender, EventArgs e)
```

Comment:
Unrelated, I wrote a local function the other day and even in VS with proper syntax highlighting I still kinda want to add `// ------------ end of method -------------` just above them :)

Reply:
I generally like to add a redundant `return` in these cases. Missed this one, thanks. Added.